filemanager.go

```diff
@@ -0,0 +1,350 @@
+// Copyright 2023 qbee.io
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package client
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// Manager manages the sync operation.
+type Manager struct {
+	// client is the filemanager client.
+	client *Client
+	// deleteMissing deletes files on the remote that are not present locally.
+	deleteMissing bool
+	// dryrun does not perform any changes.
+	dryrun bool
+	// remoteFiles is a map of the remote files.
+	remoteFiles map[string]File
+	// localFiles is a map of the local files.
+	localFiles map[string]File
+}
+
+// fileInfo represents a file in the filemanager.
+type fileInfo struct {
+	File
+}
+
+// NewFileManager returns a new filemanager.
+func NewFileManager() *Manager {
+	return &Manager{
+		remoteFiles: make(map[string]File),
+		localFiles:  make(map[string]File),
+	}
+}
+
+// WithClient sets the client
+func (m *Manager) WithClient(client *Client) *Manager {
+	m.client = client
+	return m
+}
+
+// WithDelete sets the delete flag
+func (m *Manager) WithDelete(del bool) *Manager {
+	m.deleteMissing = del
+	return m
+}
+
+// WithDryRun sets the dryrun flag
+func (m *Manager) WithDryRun(dryrun bool) *Manager {
+	m.dryrun = dryrun
+	return m
+}
+
+// Sync synchronizes the local directory with the filemanager.
+func (m *Manager) Sync(ctx context.Context, source, dest string) error {
+
+	if err := m.snapshotLocal(source); err != nil {
+		return err
+	}
+
+	if err := m.snapshotRemote(ctx, dest); err != nil {
+		return err
+	}
+
+	updates, err := m.filterUploads()
+
+	if err != nil {
+		return err
+	}
+
+	for _, uploadFile := range updates {
+		if uploadFile.IsDir {
+			continue
+		}
+
+		if err := m.upload(ctx, uploadFile, source, dest); err != nil {
+			return err
+		}
+	}
+
+	if m.deleteMissing {
+		// remove all files that from remote that are present locally
+		for remoteRelativeName := range m.remoteFiles {
+			if _, ok := m.localFiles[remoteRelativeName]; ok {
+				delete(m.remoteFiles, remoteRelativeName)
+			}
+		}
+		// delete all remaining files
+		return m.deleteRemoteRecursive()
+	}
+
+	return nil
+}
+
+// Remove deletes all files in the given path.
```

Comment:
`deletes` is redundant. Can be just `Remove all files in the given path.`

Reply:
Also, more details could be provided regarding the used parameters, e.g.:

```
// Remove all files under provided remotePath.
//
// If recursive flag is set to true and remotePath points to a directory,
// this method will remove all nested files and directories recursively.
// Otherwise, it will return an error when deleting a non-empty directory.
```